File: app/src/main/java/com/drdisagree/iconify/xposed/mods/IconUpdater.java
package com.drdisagree.iconify.xposed.mods;

/* Modified from AOSPMods
 * https://github.com/siavash79/AOSPMods/blob/canary/app/src/main/java/sh/siava/AOSPMods/launcher/AOSPModsIconUpdater.java
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

import static de.robv.android.xposed.XposedBridge.hookAllConstructors;
import static de.robv.android.xposed.XposedBridge.hookAllMethods;
import static de.robv.android.xposed.XposedBridge.log;
import static de.robv.android.xposed.XposedHelpers.callMethod;
import static de.robv.android.xposed.XposedHelpers.findClass;

import android.content.Context;
import android.os.UserHandle;

import com.drdisagree.iconify.BuildConfig;
import com.drdisagree.iconify.xposed.ModPack;

import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

public class IconUpdater extends ModPack implements IXposedHookLoadPackage {

    private static final String TAG = "Iconify - " + IconUpdater.class.getSimpleName() + ": ";
    private Object LauncherModel = null;

    public IconUpdater(Context context) {
        super(context);
    }

    @Override
    public void updatePrefs(String... Key) {
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        try {
            Class<?> LauncherModelClass = findClass("com.android.launcher3.LauncherModel", lpparam.classLoader);
            Class<?> BaseDraggingActivityClass = findClass("com.android.launcher3.BaseDraggingActivity", lpparam.classLoader);

            hookAllConstructors(LauncherModelClass, new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) {
                    LauncherModel = param.thisObject;
                }
            });

            hookAllMethods(BaseDraggingActivityClass, "onResume", new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) {
                    try {
                        if (LauncherModel != null) {
                            callMethod(LauncherModel, "onAppIconChanged", BuildConfig.APPLICATION_ID, UserHandle.getUserHandleForUid(0));
                        }
                    } catch (Throwable throwable) {
                        log(TAG + throwable);
                    }
                }
            });
        } catch (Throwable ignored) {
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/xposed/modules/IconUpdater.kt
package com.drdisagree.iconify.xposed.modules

import android.content.Context
import android.os.UserHandle
import com.drdisagree.iconify.BuildConfig
import com.drdisagree.iconify.xposed.ModPack
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedBridge.hookAllConstructors
import de.robv.android.xposed.XposedBridge.hookAllMethods
import de.robv.android.xposed.XposedBridge.log
import de.robv.android.xposed.XposedHelpers.callMethod
import de.robv.android.xposed.XposedHelpers.findClass
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam

class IconUpdater(context: Context?) : ModPack(context!!) {

    private var launcherModel: Any? = null

    override fun updatePrefs(vararg key: String) {}

    override fun handleLoadPackage(loadPackageParam: LoadPackageParam) {
        try {
            val launcherModelClass = findClass(
                "com.android.launcher3.LauncherModel",
                loadPackageParam.classLoader
            )
            val baseDraggingActivityClass = findClass(
                "com.android.launcher3.BaseDraggingActivity",
                loadPackageParam.classLoader
            )

            hookAllConstructors(launcherModelClass, object : XC_MethodHook() {
                override fun afterHookedMethod(param: MethodHookParam) {
                    launcherModel = param.thisObject
                }
            })

            hookAllMethods(baseDraggingActivityClass, "onResume", object : XC_MethodHook() {
                override fun afterHookedMethod(param: MethodHookParam) {
                    try {
                        if (launcherModel != null) {
                            callMethod(
                                launcherModel,
                                "onAppIconChanged",
                                BuildConfig.APPLICATION_ID,
                                UserHandle.getUserHandleForUid(0)
                            )
                        }
                    } catch (throwable: Throwable) {
                        log(TAG + throwable)
                    }
                }
            })
        } catch (ignored: Throwable) {
        }
    }

    companion object {
        private val TAG = "Iconify - ${IconUpdater::class.java.simpleName}: "
    }
}